barretenberg/cpp/src/barretenberg/bb/api_client_ivc.hpp

```diff
@@ -0,0 +1,266 @@
+{
+    std::ifstream fin;
+    fin.open(filename, std::ios::ate | std::ios::binary);
+    if (!fin.is_open()) {
+        throw std::invalid_argument("file not found");
+    }
+    if (fin.tellg() == -1) {
+        throw std::invalid_argument("something went wrong");
+    }
+
+    uint64_t fsize = static_cast<uint64_t>(fin.tellg());
+    fin.seekg(0, std::ios_base::beg);
+
+    T result;
+    char* encoded_data = new char[fsize];
+    fin.read(encoded_data, static_cast<std::streamsize>(fsize));
+    msgpack::unpack(encoded_data, fsize).get().convert(result);
+    return result;
+}
+
+// TODO(#7371) find a home for this
+acir_format::WitnessVector witness_map_to_witness_vector(std::map<std::string, std::string> const& witness_map)
+{
+    acir_format::WitnessVector wv;
+    size_t index = 0;
+    for (auto& e : witness_map) {
+        uint64_t value = std::stoull(e.first);
+        // ACIR uses a sparse format for WitnessMap where unused witness indices may be left unassigned.
+        // To ensure that witnesses sit at the correct indices in the `WitnessVector`, we fill any indices
+        // which do not exist within the `WitnessMap` with the dummy value of zero.
+        while (index < value) {
+            wv.push_back(fr(0));
+            index++;
+        }
+        wv.push_back(fr(uint256_t(e.second)));
+        index++;
+    }
+    return wv;
+}
+
+std::vector<uint8_t> decompress(uint8_t* bytes, size_t size)
+{
+    std::vector<uint8_t> content;
+    // initial size guess
+    content.resize(1024ULL * 128ULL);
+    for (;;) {
+        auto decompressor = std::unique_ptr<libdeflate_decompressor, void (*)(libdeflate_decompressor*)>{
+            libdeflate_alloc_decompressor(), libdeflate_free_decompressor
+        };
+        size_t actual_size = 0;
+        libdeflate_result decompress_result = libdeflate_gzip_decompress(
+            decompressor.get(), bytes, size, std::data(content), std::size(content), &actual_size);
+        if (decompress_result == LIBDEFLATE_INSUFFICIENT_SPACE) {
+            // need a bigger buffer
+            content.resize(content.size() * 2);
+            continue;
+        }
+        if (decompress_result == LIBDEFLATE_BAD_DATA) {
+            throw std::invalid_argument("bad gzip data in bb main");
+        }
+        content.resize(actual_size);
+        break;
+    }
+    return content;
+}
+
+class ClientIVCAPI : public API {
+    static std::vector<acir_format::AcirProgram> _build_folding_stack(const std::string& input_type,
+                                                                      const std::filesystem::path& bytecode_path,
+                                                                      const std::filesystem::path& witness_path)
+    {
+        using namespace acir_format;
+
+        std::vector<AcirProgram> folding_stack;
+
+        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1162): Efficiently unify ACIR stack parsing
```

Comment:
nit: we should uniformly either use - or _ in input type

Reply:
I think I _did_ achieve this, but I see that some of the comments about deprecation weren't updated to refer to the _ version of things. I suspect this is what you're referring to?